src/clasp_facade.cpp

```diff
@@ -555,8 +555,214 @@ double _getResult(const SolveResult* r) { return static_cast<double>(r->operator
 double _getSignal(const SolveResult* r) { return static_cast<double>(r->signal); }
 double _getExhausted(const SolveResult* r) { return static_cast<double>(r->exhausted()); }
 }
+
+ClaspFacade::UserdefinedStats::UserdefinedStats() {
+	maps_.emplace_back();
+	root_ = maps_.size()-1;
+	mapper_.emplace_back(root_, Potassco::Statistics_t::Map);
+}
+
+const char* ClaspFacade::UserdefinedStats::name() const {
+	return "userdefined";
+}
+
+size_t ClaspFacade::UserdefinedStats::root() const {
+	return root_;
+}
+
+bool ClaspFacade::UserdefinedStats::exists(size_t key, const std::string& name, size_t& target) const {
+	validKey(key);
+	if (mapper_[key].second != Potassco::Statistics_t::Map) {
+		throw std::logic_error("Statistic function can only be applied to a map");
+	}
+	auto v = connections_.find(key);
+	if (v == connections_.end()) { return false; }
+	for (const auto& i : v->second) {
+		if (strings_[i.index] == name) {
+		  target = i.target;
+		  return true;
+		}
+	}
+	return false;
+}
+
+bool ClaspFacade::UserdefinedStats::exists(size_t key, const std::string& name, Potassco::Statistics_t type, size_t& target) const {
+	validKey(key);
+	if (mapper_[key].second != Potassco::Statistics_t::Map) {
+		throw std::logic_error("Statistic function can only be applied to a map");
+	}
+	auto v = connections_.find(key);
+	if (v == connections_.end()) { return false; }
+	for (const auto& i : v->second) {
+		if (strings_[i.index] == name) {
+			target = i.target;
+			return mapper_[i.target].second == type;
+		}
+	}
+	return false;
+}
+
+bool ClaspFacade::UserdefinedStats::exists(size_t key, size_t index, size_t& target) const {
+	validKey(key);
+	if (mapper_[key].second != Potassco::Statistics_t::Array) {
+		throw std::logic_error("Statistic function can only be applied to an array");
+	}
+	auto v = connections_.find(key);
+	if (v == connections_.end()) { return false; }
+	for (const auto& i : v->second) {
```

Comment:
Are sparse arrays really needed? The existing stats arrays all behave like normal c++ arrays, i.e. all elements in the range [0, size) are considered valid.